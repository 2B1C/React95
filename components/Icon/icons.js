import bat from '../shared/assets/icons/bat.ico';
import bat_exec from '../shared/assets/icons/bat_exec.ico';
import bat_exec_2 from '../shared/assets/icons/bat_exec_2.ico';
import bat_wait from '../shared/assets/icons/bat_wait.ico';
import computer from '../shared/assets/icons/computer.ico';
import computer_2 from '../shared/assets/icons/computer_2.ico';
import computer_3 from '../shared/assets/icons/computer_3.ico';
import computer_4 from '../shared/assets/icons/computer_4.ico';
import computer_find from '../shared/assets/icons/computer_find.ico';
import defrag from '../shared/assets/icons/defrag.ico';
import desktop from '../shared/assets/icons/desktop.ico';
import file_delete from '../shared/assets/icons/file_delete.ico';
import file_find from '../shared/assets/icons/file_find.ico';
import file_find2 from '../shared/assets/icons/file_find2.ico';
import file_corrupted from '../shared/assets/icons/file_corrupted.ico';
import file_font from '../shared/assets/icons/file_font.ico';
import file_font_2 from '../shared/assets/icons/file_font_2.ico';
import file_settings from '../shared/assets/icons/file_settings.ico';
import file_text from '../shared/assets/icons/file_text.ico';
import file_text_settings from '../shared/assets/icons/file_text_settings.ico';
import file_transfer from '../shared/assets/icons/file_transfer.ico';
import files from '../shared/assets/icons/files.ico';
import folder from '../shared/assets/icons/folder.ico';
import folder_exe from '../shared/assets/icons/folder_exe.ico';
import folder_exe2 from '../shared/assets/icons/folder_exe2.ico';
import folder_file from '../shared/assets/icons/folder_file.ico';
import folder_font from '../shared/assets/icons/folder_font.ico';
import folder_open from '../shared/assets/icons/folder_open.ico';
import folder_print from '../shared/assets/icons/folder_print.ico';
import folder_rename from '../shared/assets/icons/folder_rename.ico';
import folder_settings from '../shared/assets/icons/folder_settings.ico';
import folder_settings_2 from '../shared/assets/icons/folder_settings_2.ico';
import folder_shared from '../shared/assets/icons/folder_shared.ico';
import globe from '../shared/assets/icons/globe.ico';
import hand from '../shared/assets/icons/hand.ico';
import help_book from '../shared/assets/icons/help_book.ico';
import loader_bat from '../shared/assets/icons/loader_bat.ico';
import logo from '../shared/assets/icons/logo.ico';
import media_cd from '../shared/assets/icons/media_cd.ico';
import memory from '../shared/assets/icons/memory.ico';
import network from '../shared/assets/icons/network.ico';
import network_2 from '../shared/assets/icons/network_2.ico';
import network_3 from '../shared/assets/icons/network_3.ico';
import printer from '../shared/assets/icons/printer.ico';
import printer_calendar from '../shared/assets/icons/printer_calendar.ico';
import printer_drive from '../shared/assets/icons/printer_drive.ico';
import printer_shared from '../shared/assets/icons/printer_shared.ico';
import reader_cd from '../shared/assets/icons/reader_cd.ico';
import reader_cd_2 from '../shared/assets/icons/reader_cd_2.ico';
import reader_closed from '../shared/assets/icons/reader_closed.ico';
import reader_disket from '../shared/assets/icons/reader_disket.ico';
import reader_disket_casset from '../shared/assets/icons/reader_disket_casset.ico';
import reader_disket2 from '../shared/assets/icons/reader_disket2.ico';
import reader_eject from '../shared/assets/icons/reader_eject.ico';
import reader_noshared from '../shared/assets/icons/reader_noshared.ico';
import reader_opened from '../shared/assets/icons/reader_opened.ico';
import reader_shared from '../shared/assets/icons/reader_shared.ico';
import recycle_empty from '../shared/assets/icons/recycle_empty.ico';
import recycle_file from '../shared/assets/icons/recycle_file.ico';
import recycle_filefolder from '../shared/assets/icons/recycle_filefolder.ico';
import recycle_folder from '../shared/assets/icons/recycle_folder.ico';
import recycle_full from '../shared/assets/icons/recycle_full.ico';
import settings from '../shared/assets/icons/settings.ico';
import shortcut from '../shared/assets/icons/shortcut.ico';
import shortcut2 from '../shared/assets/icons/shortcut2.ico';
import tree from '../shared/assets/icons/tree.ico';
import wordpad from '../shared/assets/icons/wordpad.ico';
import windows_explorer from '../shared/assets/icons/windows_explorer.ico';
import ms_dos from '../shared/assets/icons/ms_dos.ico';
import microsoft_network from '../shared/assets/icons/microsoft_network.ico';
import microsoft_exchange from '../shared/assets/icons/microsoft_exchange.ico';

const path = {
  bat,
  bat_exec,
  bat_exec_2,
  bat_wait,
  computer,
  computer_2,
  computer_3,
  computer_4,
  computer_find,
  defrag,
  desktop,
  file_delete,
  file_find,
  file_find2,
  file_corrupted,
  file_font,
  file_font_2,
  file_settings,
  file_text,
  file_text_settings,
  file_transfer,
  files,
  folder,
  folder_exe,
  folder_exe2,
  folder_file,
  folder_font,
  folder_open,
  folder_print,
  folder_rename,
  folder_settings,
  folder_settings_2,
  folder_shared,
  globe,
  hand,
  help_book,
  loader_bat,
  logo,
  media_cd,
  memory,
  network,
  network_2,
  network_3,
  printer,
  printer_calendar,
  printer_drive,
  printer_shared,
  reader_cd,
  reader_cd_2,
  reader_closed,
  reader_disket,
  reader_disket_casset,
  reader_disket2,
  reader_eject,
  reader_noshared,
  reader_opened,
  reader_shared,
  recycle_empty,
  recycle_file,
  recycle_filefolder,
  recycle_folder,
  recycle_full,
  settings,
  shortcut,
  shortcut2,
  tree,
  wordpad,
  windows_explorer,
  ms_dos,
  microsoft_network,
  microsoft_exchange,
};

export default path;
