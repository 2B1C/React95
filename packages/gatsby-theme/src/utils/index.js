import isEmpty from './isEmpty';
import navify from './navify';

export { isEmpty, navify };
