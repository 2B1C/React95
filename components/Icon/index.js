import React from 'react';
import PropTypes from 'prop-types';
import styled, { injectGlobal } from 'styled-components';
import icons from './icons.js';

const I = styled.i`
  display: block;
  width: 30px;
  height: 32px;

  background-image: url('${({ name }) => icons[name] || 'none'}');
  background-repeat: no-repeat;
  background-position: center;
`;

class Icon extends React.Component {
  render() {
    const { name } = this.props;
    console.log(icons[name]);

    return <I name={name} />;
  }
}

Icon.propTypes = {
  name: PropTypes.string,
};

Icon.defaultProps = {
  name: null,
};

export default Icon;
