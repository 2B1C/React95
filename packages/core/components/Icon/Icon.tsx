import styled from 'styled-components';
import icons from '@react95/icons/icons.module.css';

export interface IconProps {
  name: keyof typeof icons;
}

const Icon = styled.i
  .withConfig({
    shouldForwardProp: (prop, defaultValidatorFn) =>
      !['name'].includes(prop) && defaultValidatorFn(prop),
  })
  .attrs<IconProps>(({ name }) => ({
    className: name,
  }))<IconProps>`
  display: block;
`;

export default Icon;
