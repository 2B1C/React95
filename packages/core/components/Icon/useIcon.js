import { useState, useEffect } from 'react';
import icons from '@react95/icons';
import ico from 'icojs';

const MIME_TYPE = 'image/png';

function icoParse(file) {
  if (ico.isICO(file)) {
    return ico.parse(file);
  }

  return [];
}

const useIcon = ({ name, size, fallback }) => {
  const [data, setAvailableIcons] = useState({
    iconUrl: '',
    availableIcons: [{}],
  });

  const changeIconUrl = newUrl => {
    setAvailableIcons(iconData => ({ ...iconData, iconUrl: newUrl }));
  };

  useEffect(() => {
    async function fetchIcon() {
      const response = await fetch(icons[name]);
      const iconBuffer = await response.arrayBuffer();

      const allIcons = await icoParse(iconBuffer);

      const iconsToRender = allIcons.map(i => ({
        size: i.width,
        url: URL.createObjectURL(new Blob([i.buffer], { type: MIME_TYPE })),
        bit: i.bbt,
      }));

      let url;
      if (fallback) {
        url = iconsToRender[0].url;
      } else {
        const match = iconsToRender.find(i => i.size === size);
        url = match ? match.url : undefined;
      }

      setAvailableIcons({ iconUrl: url, availableIcons: iconsToRender });
    }

    fetchIcon();
  }, []);

  return {
    changeIconUrl,
    iconUrl: data.iconUrl,
    availableIcons: data.availableIcons,
  };
};

export default useIcon;
