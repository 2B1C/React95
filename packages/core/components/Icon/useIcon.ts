import * as React from 'react';
import icons from '@react95/icons';
import ico from 'icojs';

import 'isomorphic-fetch';

import { IconProps } from './Icon';

const MIME_TYPE = 'image/png';

async function icoParse(file: ArrayBuffer): Promise<ico.ParsedImage[]> {
  if (ico.isICO(file)) {
    return ico.parse(file);
  }

  return [];
}

export interface IAvailableIcon {
  size: number;
  url: string;
  bpp: number;
}

type UseIconState = {
  iconUrl?: string;
  availableIcons: Array<IAvailableIcon>;
};

const useIcon = ({
  name,
  size,
  fallback,
  bpp = 4,
  variant = 1,
}: Partial<IconProps>) => {
  const [data, setAvailableIcons] = React.useState<UseIconState>({
    iconUrl: '',
    availableIcons: [],
  });

  const changeIconUrl = (newUrl: string) => {
    setAvailableIcons(iconData => ({ ...iconData, iconUrl: newUrl }));
  };

  React.useEffect(() => {
    async function fetchIcon() {
      const response = await fetch(icons[name]);
      const iconBuffer = await response.arrayBuffer();

      const allIcons = await icoParse(iconBuffer);

      const iconsToRender = allIcons.map(i => ({
        size: i.width,
        url: URL.createObjectURL(new Blob([i.buffer], { type: MIME_TYPE })),
        bpp: i.bpp,
      }));

      let url = undefined;
      const match = iconsToRender.filter(i => i.size === size && i.bpp === bpp);

      if (match.length > 0) {
        const found = match[variant - 1];
        url = found.url;
      } else if (fallback) {
        url = iconsToRender[0].url;
      }

      setAvailableIcons({ iconUrl: url, availableIcons: iconsToRender });
    }

    fetchIcon();
  }, []);

  return {
    changeIconUrl,
    iconUrl: data.iconUrl,
    availableIcons: data.availableIcons,
  };
};

export default useIcon;
