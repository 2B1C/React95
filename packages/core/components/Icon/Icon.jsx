import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

import useIcon from './useIcon';

const I = styled.i`
  display: block;

  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;

  ${({ width, height, url }) => css`
    width: ${width}px;
    height: ${height}px;
    background-image: ${url ? `url('${url}')` : 'none'};
  `}
`;

const Icon = ({ name, width, height, size, fallback, ...rest }) => {
  const { iconUrl, changeIconUrl, availableIcons } = useIcon({
    name,
    size,
    fallback,
  });

  useEffect(() => {
    const icon = availableIcons.find(i => i.size === size);

    changeIconUrl(icon ? icon.url : undefined);
  }, [size]);

  return (
    <I
      name={name}
      width={width || size}
      height={height || size}
      {...rest}
      url={iconUrl}
    />
  );
};

Icon.propTypes = {
  name: PropTypes.string,
  width: PropTypes.number,
  height: PropTypes.number,
  fallback: PropTypes.bool,
  size: PropTypes.number,
};

Icon.defaultProps = {
  name: null,
  width: undefined,
  height: undefined,
  fallback: true,
  size: 32,
};

export default Icon;
