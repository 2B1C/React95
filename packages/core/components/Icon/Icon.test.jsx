import React from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { render, act } from '../shared/test/utils';

import useIcon from './useIcon';
import Icon from './Icon';

const BUFFER_SIZES = {
  16: 'buffer16',
  32: 'buffer32',
};

jest.mock('icojs', () => ({
  isICO: jest.fn(() => true),
  parse: jest.fn(() =>
    Promise.resolve([
      {
        width: 16,
        buffer: 'buffer16',
        bbt: 4,
      },
      {
        width: 32,
        buffer: 'buffer32',
        bbt: 4,
      },
    ]),
  ),
}));

beforeAll(() => {
  global.fetch = jest.fn().mockImplementation(() =>
    Promise.resolve({
      arrayBuffer: jest.fn(() => ({})),
    }),
  );

  global.Blob = class Blob {
    constructor(buff) {
      this.buffer = buff;
    }

    toString() {
      return this.buffer.toString();
    }
  };

  global.URL.createObjectURL = jest.fn((data) => data.toString());
});

describe('<Icon />', () => {
  it('should match snapshot', async () => {
    const { container } = render(<Icon name="bat" />);

    await act(async () => {
      await Promise.resolve(container);
    });

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with different width and height', async () => {
    const { container } = render(
      <>
        <Icon name="bat" width={10} height={10} />
        <Icon name="bat" width={20} height={20} />
        <Icon name="bat" width={30} height={30} />
      </>,
    );

    await act(async () => {
      await Promise.resolve(container);
    });

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with different size', async () => {
    const { container } = render(<Icon name="bat" size={16} />);

    await act(async () => {
      await Promise.resolve(container);
    });

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with different size and without a fallback', async () => {
    const { container } = render(
      <Icon name="bat" size={64} fallback={false} />,
    );

    await act(async () => {
      await Promise.resolve(container);
    });

    expect(container).toMatchSnapshot();
  });
});

describe('useIcon', () => {
  it('should return the icon by given a name', async () => {
    const { result, waitForNextUpdate } = renderHook(
      ({ name, size }) => useIcon({ name, size }),
      {
        initialProps: { name: 'bat', size: 16 },
      },
    );

    expect(result.current.iconUrl).toBe('');

    await waitForNextUpdate();

    expect(result.current.iconUrl).toBe(BUFFER_SIZES[16]);

    expect(result.current.availableIcons).toHaveLength(
      Object.keys(BUFFER_SIZES).length,
    );
  });

  it('should return no iconUrl if you ommit the size', async () => {
    const { result, waitForNextUpdate } = renderHook(
      ({ name, size }) => useIcon({ name, size }),
      {
        initialProps: { name: 'bat' },
      },
    );

    expect(result.current.iconUrl).toBe('');

    await waitForNextUpdate();

    expect(result.current.iconUrl).toBe(undefined);

    expect(result.current.availableIcons).toHaveLength(
      Object.keys(BUFFER_SIZES).length,
    );
  });

  it('should return a fallback iconUrl by given a wrong size', async () => {
    const { result, waitForNextUpdate } = renderHook(
      ({ name, size, fallback }) => useIcon({ name, size, fallback }),
      {
        initialProps: { name: 'bat', size: 48, fallback: true },
      },
    );

    expect(result.current.iconUrl).toBe('');

    await waitForNextUpdate();

    expect(result.current.iconUrl).toBe(Object.values(BUFFER_SIZES)[0]);

    expect(result.current.availableIcons).toHaveLength(
      Object.keys(BUFFER_SIZES).length,
    );
  });
});
