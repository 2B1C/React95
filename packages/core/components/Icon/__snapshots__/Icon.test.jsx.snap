// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon /> should match snapshot 1`] = `
.c0 {
  display: block;
  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;
  width: 32px;
  height: 32px;
  background-image: url('buffer32');
}

<div>
  <i
    class="c0"
    height="32"
    name="bat"
    width="32"
  />
</div>
`;

exports[`<Icon /> should match snapshot with different size 1`] = `
.c0 {
  display: block;
  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;
  width: 16px;
  height: 16px;
  background-image: url('buffer16');
}

<div>
  <i
    class="c0"
    height="16"
    name="bat"
    width="16"
  />
</div>
`;

exports[`<Icon /> should match snapshot with different size and without a fallback 1`] = `
.c0 {
  display: block;
  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;
  width: 64px;
  height: 64px;
  background-image: none;
}

<div>
  <i
    class="c0"
    height="64"
    name="bat"
    width="64"
  />
</div>
`;

exports[`<Icon /> should match snapshot with different width and height 1`] = `
.c0 {
  display: block;
  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;
  width: 10px;
  height: 10px;
  background-image: url('buffer32');
}

.c1 {
  display: block;
  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;
  width: 20px;
  height: 20px;
  background-image: url('buffer32');
}

.c2 {
  display: block;
  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;
  width: 30px;
  height: 30px;
  background-image: url('buffer32');
}

<div>
  <i
    class="c0"
    height="10"
    name="bat"
    width="10"
  />
  <i
    class="c1"
    height="20"
    name="bat"
    width="20"
  />
  <i
    class="c2"
    height="30"
    name="bat"
    width="30"
  />
</div>
`;
